core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoticeSchemaGenerator.java

```diff
@@ -0,0 +1,263 @@
+   *       "attributionId": {
+   *         "type": "string"
+   *       },
+   *       "csvRowNumber": {
+   *         "type": "integer"
+   *       }
+   *     }
+   *   },
+   *   "duplicate_fare_rule_zone_id_fields": {
+   *     "type": "object",
+   *     "properties": {
+   *       "csvRowNumber": {
+   *         "type": "integer"
+   *       },
+   *       "fareId": {
+   *         "type": "string"
+   *       },
+   *       "previousCsvRowNumber": {
+   *         "type": "integer"
+   *       },
+   *       "previousFareId": {
+   *         "type": "string"
+   *       }
+   *     }
+   *   },
+   *   ...
+   * }
+   * }</pre>
+   *
+   * @param packages List of packages where notices are declared. Use {@link
+   *     #DEFAULT_NOTICE_PACKAGES} and add your custom Java packages here, if any
+   * @return a {@link JsonObject} describing all notices in given packages (see above)
+   * @throws IOException
+   */
+  public static JsonObject jsonSchemaForPackages(List<String> packages) throws IOException {
+    JsonObject schema = new JsonObject();
+    for (Map.Entry<String, Map<String, Class<?>>> entry :
+        contextFieldsInPackages(packages).entrySet()) {
+      schema.add(
+          Notice.getCode(entry.getKey()), jsonSchemaForNotice(entry.getKey(), entry.getValue()));
+    }
+    return schema;
+  }
+
+  /**
+   * Convenient function to find all notices in given packages and describe their fields.
+   *
+   * <p>The returned map has looks this way:
+   *
+   * <pre>{@code
+   * {
+   *   "AttributionWithoutRoleNotice": {
+   *     "attributionId": String.class,
+   *     "csvRowNumber": Long.class,
+   *   }
+   * }
+   * }</pre>
+   *
+   * @param packages List of packages where notices are declared
+   * @return a map describing all notices in given packages (see above)
+   * @throws IOException
+   */
+  @VisibleForTesting
+  static Map<String, Map<String, Class<?>>> contextFieldsInPackages(List<String> packages)
+      throws IOException {
+    // Return a sorted TreeMap for stable results.
+    Map<String, Map<String, Class<?>>> contextFieldsByNotice = new TreeMap<>();
+    for (Class<Notice> noticeClass : findNoticeSubclasses(packages)) {
+      contextFieldsByNotice.put(noticeClass.getSimpleName(), contextFieldsForNotice(noticeClass));
+    }
+
+    return contextFieldsByNotice;
+  }
+
+  @VisibleForTesting
+  static Map<String, Class<?>> contextFieldsForNotice(Class<? extends Notice> noticeClass) {
+    // Return a sorted TreeMap for stable results.
+    Map<String, Class<?>> fields = new TreeMap<>();
+    for (Field field : noticeClass.getDeclaredFields()) {
+      fields.put(field.getName(), field.getType());
+    }
+    return fields;
+  }
+
+  private static boolean isSubclassOf(Class<?> parent, Class<?> child) {
+    return !child.equals(parent) && parent.isAssignableFrom(child);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static void maybeAddNoticeClass(Class<?> clazz, List<Class<Notice>> notices) {
```

Comment:
For clarity:

```suggestion
  private static void addNestedNoticeClass(Class<?> clazz, List<Class<Notice>> notices) {
```

Reply:
No, it is not necessarily nested. It is called twice: first time for top-level classes, second for nested ones.